sway-lib-core/src/raw_ptr.sw

```diff
@@ -0,0 +1,67 @@
+library raw_ptr;
+
+impl raw_ptr {
+    /// Returns `true` if the pointer is null.
+    pub fn is_null(self) -> bool {
+        let addr = asm(ptr: self) { ptr: u64 };
+        __eq(addr, 0)
+    }
+
+    /// Gets the address of the pointer.
+    pub fn addr(self) -> u64 {
+        asm(ptr: self) { ptr: u64 }
+    }
```

Comment:
How about we be more explicit here and call this `as_u64` or something like that? `addr()` seems like something you'd apply on a variable to get its address, and not on a pointer.

Reply:
Pointers can have more than an address so it made sense to me. Are we going to keep `raw_ptr`, or will it eventually turn into a typed pointer where it has a length too?

I was thinking:
- Turn `raw_ptr` into `ptr<T>` (or `ptr T` or `&T`).
- Add slices (`ptr<[T]>`?) so pointers can have a length in runtime which can be used for serialization.
- Extend the spec so our serialization format supports dynamic lengths.
- Implement `Serialize/Deserialize` traits for primitives and slices.
- Implement `Into<ptr<[T]>> for Vec<T>`.
- Implicitly call `Serialize::serialize` and `Deserialize::deserialize` when reading `script_data` or returning from a script. (Only allow `Into<Serialize>` to be returned, `Into<Deserialize>` to be inputted.)

Notes:
- Might need https://github.com/FuelLabs/sway/issues/3018
- Related: https://github.com/FuelLabs/sway/issues/2449

Reply:
I'm not sure what the future of `raw_ptr` is going to be and your suggestions could certainly be possible! (Worth documenting in an issue for bikeshedding).

> Pointers can have more than an address so it made sense to me.

What does that mean specifically? At any given point in execution, a `raw_ptr` will have a single `u64` value. When I read `addr(ptr)`, in my head this is "address of `ptr`" while in fact all we're doing is an explicit conversion of a pointer to a `u64`, hence the name I suggested.

Reply:
ATM yes but they could have metadata: https://doc.rust-lang.org/std/ptr/trait.Pointee.html#pointer-metadata

Reply:
> * Turn `raw_ptr` into `ptr<T>` (or `ptr T` or `*T`).

We need to be careful and ensure users can't _easily_ get the address of variables and/or create hacks for accessing data they shouldn't really be allowed to.  (Hard to ensure while we have ASM blocks, but still...)  It'd be good to try and keep devs from thinking in terms of pointers as much as possible.

Sway isn't a 100% recreation of Rust for the blockchain, even though it looks that way -- we need to keep it as safe as possible for writing robust contracts without needing a borrow checker.

Reply:
Yeah I'd imagine only library authors should ever think about pointers.

Reply:
We actually don't need `addr()`. I've removed it.